llpc/context/llpcCompiler.cpp

```diff
@@ -571,6 +572,253 @@ template <typename T> static FormattedBytes formatBytesLittleEndian(ArrayRef<T>
+  if (context->isGraphics()) {
+    auto pipelineInfo = reinterpret_cast<const GraphicsPipelineBuildInfo *>(context->getPipelineBuildInfo());
+    cacheHash = PipelineDumper::generateHashForGraphicsPipeline(pipelineInfo, true, true, stage);
+  } else {
+    auto pipelineInfo = reinterpret_cast<const ComputePipelineBuildInfo *>(context->getPipelineBuildInfo());
+    cacheHash = PipelineDumper::generateHashForComputePipeline(pipelineInfo, true, true);
+  }
+  // Note that this code updates m_pipelineHash of the pipeline context. It
+  // must be restored before we link the pipeline ELF at the end of this for-loop.
+  context->getPipelineContext()->setHashForCacheLookUp(cacheHash);
+  LLPC_OUTS("Finalized hash for " << getUnlinkedShaderStageName(stage) << " stage cache lookup: "
+                                  << format_hex(context->getPipelineContext()->get128BitCacheHashCode()[0], 18) << ' '
+                                  << format_hex(context->getPipelineContext()->get128BitCacheHashCode()[1], 18)
+                                  << '\n');
+
+  Result result = Result::Success;
+  CacheAccessor cacheAccessor(context, cacheHash, getInternalCaches());
+  if (cacheAccessor.isInCache()) {
+    BinaryData elfBin = cacheAccessor.getElfFromCache();
+    auto data = reinterpret_cast<const char *>(elfBin.pCode);
+    elfPackage.assign(data, data + elfBin.codeSize);
+    LLPC_OUTS("Cache hit for shader stage " << getUnlinkedShaderStageName(stage) << "\n");
+    for (ShaderStage gfxStage : shaderStages)
+      stageCacheAccesses[gfxStage] =
+          cacheAccessor.hitInternalCache() ? CacheAccessInfo::InternalCacheHit : CacheAccessInfo::CacheHit;
+  } else {
+    LLPC_OUTS("Cache miss for shader stage " << getUnlinkedShaderStageName(stage) << "\n");
+    for (ShaderStage gfxStage : shaderStages)
+      stageCacheAccesses[gfxStage] = CacheAccessInfo::CacheMiss;
+
+    // There was a cache miss, so we need to build the relocatable shader for
+    // this stage.
+    const PipelineShaderInfo *singleStageShaderInfo[ShaderStageNativeStageCount] = {nullptr, nullptr, nullptr,
+                                                                                    nullptr, nullptr, nullptr};
+
+    for (ShaderStage nativeStage : shaderStages)
+      singleStageShaderInfo[nativeStage] = shaderInfo[nativeStage];
+
+    result = buildPipelineInternal(context, singleStageShaderInfo, PipelineLink::Unlinked, nullptr, &elfPackage,
+                                   stageCacheAccesses);
+    if (result == Result::Success) {
+      // Add the result to the cache.
+      BinaryData elfBin = {elfPackage.size(), elfPackage.data()};
+      cacheAccessor.setElfInCache(elfBin);
+      LLPC_OUTS("Updating the cache for unlinked shader stage " << getUnlinkedShaderStageName(stage) << "\n");
+    }
+  }
+  context->getPipelineContext()->setShaderStageMask(originalShaderStageMask);
+  context->getPipelineContext()->setHashForCacheLookUp(originalCacheHash);
```

Comment:
Since we are refactoring this code.  Could you put this into an `make_scope_exit` like you did to release the context above?